### ISSUE TITLE
[4.x] waitTask returns undefined

### ISSUE THREAD

**Issue by spitlo** [algolia/algoliasearch-client-javascript#1030]
While migrating my app from v3 to v4 I noticed that `waitTask` no longer returns `'published'`, but rather `undefined`. I’m using it the old way, like this:
```
const task = index.waitTask(taskID)
```
Is this still supported? The 4.x docs says to use `.wait()` but I can see no difference in response if I use that or not, and it doesn’t seem to wait for the tasks.
When I do: 
```
const response = await index.saveObjects(objects)
```
The response is:
```
{ taskIDs: [ 00000000000 ], objectIDs: [ '13', '14', '15' ] }
```
If I instead do: 
```
const response = await index.saveObjects(objects).wait()
```
The response is still:
```
{ taskIDs: [ 00000000000 ], objectIDs: [ '13', '14', '15' ] }
```

Perhaps I’m missing something?

On a related note (Tell me if this should be a separate issue), the return I’m seeing from `index.saveObjects` doesn’t match the 4.x documentation. According to the 4.x docs, the response should have `taskID` in it, but as you can see above, the response I’m getting is an array, `taskIDs`, containing a singe `taskID`. Looking at the code, `index.saveObjects` seems to respond with a `chunkedBatchResponse`, but looking at the closest match I can find in the documentation for that, [Custom Batch](https://www.algolia.com/doc/api-reference/api-methods/batch/), the response should be an object.

Is it the behaviour or the documentation that is faulty?

**Comment by nunomaduro** [comment]
Thanks for creating the issue.

All write operations in Algolia are asynchronous by design. When you perform a `await index.saveObjects(objects)`, the operation gets sent in Algolia, but takes some milliseconds to actually start indexing in your data.

```
// Publish a save objects operation
const response = await index.saveObjects(objects);

// A search done right after may not contain your object yet.
```

If you use `.wait()`, we actually **only** resolve the original response once the indexation is done on the Algolia side.

```
// Publish a save objects operation, but only resolves the response once the indexation is complete.
const response = await index.saveObjects(objects).wait();

// A search done right after **will** contain your objects for sure
```

Concerning the response format, is also mentioned in that document, that the shape of the response may change on the api client. And that is the case of the JavaScript client. As we perform automatically batching for you, to avoid memory issues, so you get an array of `taskIDs` instead of a single `taskID`.

Let me know if you have any other question.

**Comment by spitlo** [comment]
Thanks for your quick reply!

The only thing I still don’t know is why `waitTask` returns `undefined` and not `'published'`. Should I assume that `undefined` means the task finished successfully, and expect the task to `retry()` if not? I just want to know what value to check for to verify that the task was successful.

The response I’m talking about is created [here](https://github.com/algolia/algoliasearch-client-javascript/blob/e6dec843932bdfaac13ffa1e0a11964975499078/packages/client-search/src/methods/index/waitTask.ts#L11)

On 3.x, this code:
```
const result = await index.waitTask(taskID)
console.log(`Indexing finished with result ${result}`)
```
printed "Indexing finished with result published".

On 4.x, it prints "Indexing finished with result undefined".

Thanks again for your help!

**Comment by nunomaduro** [comment]
In both `3.x` and `4.x`, when the promise of `waitTask` only gets resolved when the task is on ** published** state.

>  I just want to know what value to check for to verify that the task was successful.

You don't need todo this, normally, if you perform a `saveObjects(objects)` and you don't get any error, the task will be always completed with success in async.

If you really want to wait for the task to completed on the engine side, check perform `saveObjects(objects).wait()`
